### ISSUE TITLE
cl /kernel builds

### ISSUE THREAD

**Issue by DBJDBJ** [MicrosoftDocs/cpp-docs#2494]
In /kernel builds there is (very likely) a version of 

```cpp
void __declspec(noreturn) __stdcall
    _com_raise_error(HRESULT hr, IErrorInfo* perrinfo = 0) ;
```
Which raises the Structured Exception (of an SEH fame).  Thus `/kernel` builds "just work". Perhaps there should be some documentation on that for C++ /kernel builders.

It seems the same mechanism is implemented for MS STL  `/kernel` builds so they do "just work" too.

SEH is not C++ exceptions. C++ destructors are not called and such. I think a lot of C++ `/kernel` people might get confused because they are unaware of "SEH in action" in them builds.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: bd14af26-554d-811a-3417-4d738e85d9be
* Version Independent ID: ec58f164-e920-9204-d07d-9e6a3bce5f87
* Content: [_com_raise_error](https://docs.microsoft.com/en-us/cpp/cpp/com-raise-error?view=vs-2019)
* Content Source: [docs/cpp/com-raise-error.md](https://github.com/Microsoft/cpp-docs/blob/master/docs/cpp/com-raise-error.md)
* Product: **visual-cpp**
* Technology: **cpp-language**
* GitHub Login: @corob-msft
* Microsoft Alias: **corob**

**Comment by corob-msft** [comment]
@DBJDBJ 
I'm not sure what you're suggesting we document. You can't use `_com_raise_error` under **`/kernel`** because it throws a C++ exception. I don't see any suggestion of a **`/kernel`**-specific alternative in use in the library. You can use the same STL for kernel-mode code; it just has some differences in which member functions it exposes, controlled by a preprocessor macro. You can search the source yourself; look for the `_HAS_EXCEPTIONS` macro, defined as 0 under **`/kernel`**, which defines `_KERNEL_MODE=1`. You could conceivably raise and handle an SEH exception for COM-style errors if you wanted to in kernel code, but I'm not sure why you would. The traditional handling of HRESULTs as return values is all I can find reference to.

I'm happy to reopen the issue if you have evidence there's something happening in the Microsoft C++ compiler or libraries that I didn't see here. Depending on what you're referring to, it may be more appropriately documented in the [Windows hardware developer documentation](https://docs.microsoft.com/windows-hardware/drivers/).

**Comment by DBJDBJ** [comment]
> I'm not sure what you're suggesting we document. You can't use `_com_raise_error` under **`/kernel`** because it throws a C++ exception.

As I have written:
```cpp
void __declspec(noreturn) __stdcall
    _com_raise_error(HRESULT hr, IErrorInfo* perrinfo = 0) ;
```
Do NOT throw an C++ exception in /kernel builds. IT DOES raise SE (aka Structured Exception). 

You/me/everyone can use it. I am happily building ATL exe server using ATL **and** COM client using `<comdef.h>`. Two projects both using the /kernel build. `_com_raise_error` must exist in both of them and `_com_raise_error` must raise SE in both of them.  That is very obvious and logical?

I am/was just asking for that to be clearly documented somewhere.

**Comment by corob-msft** [comment]
@DBJDBJ 
Is the SE code EXCEPTION_RO_ORIGINATEERROR? (0x40080201L)

```cpp
// in comdef.h
inline void __declspec(noreturn) __stdcall _com_raise_error(HRESULT hr, const wchar_t* message)
{
    size_t length = (message == nullptr) ? 0 : ::wcslen(message);

    if (UINT_MAX < length)
    {
        length = 0;
    }

    ::Windows::Foundation::Diagnostics::OriginateError(hr, static_cast<unsigned int>(length), message);
    throw _com_error(hr, message);
}
```

```cpp
// in roerrorapi.h
namespace Windows
{
    namespace Foundation
    {
        namespace Diagnostics
        {
            typedef enum {
                None               = RO_ERROR_REPORTING_NONE,
                SuppressExceptions = RO_ERROR_REPORTING_SUPPRESSEXCEPTIONS,
                ForceExceptions    = RO_ERROR_REPORTING_FORCEEXCEPTIONS,
                UseSetErrorInfo    = RO_ERROR_REPORTING_USESETERRORINFO,
                SuppressSetErrorInfo = RO_ERROR_REPORTING_SUPPRESSSETERRORINFO,
            } RoErrorReportingFlags;

            DEFINE_ENUM_FLAG_OPERATORS(RoErrorReportingFlags);

            // ...

            inline BOOL OriginateError(_In_ HRESULT error, _In_ UINT cchMax,
                _When_(cchMax == 0, _In_reads_or_z_opt_(MAX_ERROR_MESSAGE_CHARS) ) 
                _When_(cchMax > 0 && cchMax < MAX_ERROR_MESSAGE_CHARS, _In_reads_or_z_(cchMax) )
                _When_(cchMax >= MAX_ERROR_MESSAGE_CHARS, _In_reads_or_z_(MAX_ERROR_MESSAGE_CHARS) ) 
                 PCWSTR message)
            {
                if (SUCCEEDED(error))
                {
                    return FALSE;
                }

                return ::RoOriginateErrorW(error, cchMax, message);
            }
            
            // ...
        }
    }
}
// ...
// The various versions of RoOriginateError will raise and catch a structured exception with an 
// exception code of EXCEPTION_RO_ORIGINATEERROR and the following information in the 
// ExceptionInformation array. The characterCount element is the number of characters in the message, 
// not including the NULL terminator.
// ExceptionInformation[0] -- HRESULT error
// ExceptionInformation[1] -- UINT characterCount
// ExceptionInformation[2] -- PCWSTR message
//
// ...
//
// As documented in MSDN, the ExceptionInformation can be retrieved in the filter expression
// of a frame-based exception handler by calling GetExceptionInformation. A debugger can
// retrieve these values by calling WaitForDebugEvent, keeping in mind that exception records have
// a different binary layout between 32bit and 64bit processes.

```

So, `_com_raise_error` calls `::Windows::Foundation::Diagnostics::OriginateError`, which invokes [RoOriginateErrorW](https://docs.microsoft.com/en-us/windows/win32/api/roerrorapi/nf-roerrorapi-rooriginateerrorw), which can raise a SE if the error reporting flags are set appropriately. I don't know how the compiler can give a pass to the `throw _com_error(hr, message);` statement under **`/kernel`**, though. It looks like the `OriginateError` call is just meant as a debugger hook. But, it's possible something similar to this code path is what you're seeing.

If there's different internal code invoked within the kernel to handle raising an error as a SE, it's opaque to me, and can't be documented here. That code doesn't belong to the C++ team. That goes for the internals of `OriginateError` as well, even though I can see them in the header.

**Comment by DBJDBJ** [comment]
Ah, we are talking now :) As you might have seen (about me here and on MS STL github repo) during last week or two I am keenly interested in C++ /kernel builds. And that (I have learned) means C++ & Structured Exceptions (SE).

Let me rephrase: MS C++ and /kernel builds == SE .Yes I know and have followed to the Ro ... . Ok and fine.  This all works and behaves quite well. 

It is just it is not documented. I mean MS STL, CL c++ behaviour under /kernel builds ... Has nothing to do with the Ro ... stuff. Just a behaviour on the surface.

I am not, but ISO C++ experts too are (it seems) unaware about all of this (example Ben Craig at https://github.com/microsoft/STL/issues/1289#issuecomment-699760934 ) ... One page on MS Docs will clarify completely.  

Basically MS STL works undert /kernel and anything else CL C++ related works under /kernel but there are no C++ exceptions in that universe. And the population at large thinks MS STL can not be used in (many) projects where exceptions are no allowed.  Neither people do know /kernel switch even exist.

**Comment by DBJDBJ** [comment]
> You can't use `_com_raise_error` under **`/kernel`** because it throws a C++ exception.

Yes, we can :)

```cpp
// (c) 2020 by dbj@dbj.org
// sampling MSVC in the /kernel build
//
#include <stdio.h>
#include <stdlib.h>
#include <comdef.h> // _com_raise_error() #221

#ifndef _KERNEL_MODE
#error This must be compiled with the /kernel switch
#endif // ! _KERNEL_MODE

#if _HAS_EXCEPTIONS
#error _HAS_EXCEPTIONS must be 0 in this build
#endif //  _HAS_EXCEPTIONS

// --------------------------------------------------------------------------
int main(int argc, char** argv)
{
    __try
    {
        __try {

            // for example, raise SE :
            // "The server did not respond after being launched."
            _com_raise_error(CO_E_SERVER_START_TIMEOUT); // comdef.h #221
            // at above point on debugger F11, debugger wants to go to
            // d:\agent\_work\3\s\src\vctools\Compiler\CxxFE\SL\vccom\comraise.cpp
            // message is
            // Exception thrown at 0x771F9962 in ... .exe: Microsoft C++ exception: _com_error at memory location 0x010FF7E0.
            // but that was not "C++ exception" thrown that was SE
        }
        __finally {
            fprintf(stderr , ":  __finally block\n");
        }
    }
    __except ( 1 /* 1 aka EXCEPTION_EXECUTE_HANDLER */ )
    {
        // "throws" are caught here
        // as now they are SE
        fprintf(stderr , ": __except block: Structured Exception caught\n");
    }
    fprintf(stderr , ": Leaving...\n");
    return 42;
} // main
```

HTH

**Comment by corob-msft** [comment]
@DBJDBJ 

After consultation with the libraries team, we think we have a handle on what's going on.

The bad news is, what you're seeing is the emergent behavior of kernel-mode code linked to the non-kernel mode COM support library. None of the functions declared in *comdef.h* are intended for or tested for use in kernel-mode code. The *vccomsup.lib* library can't even be compiled under **`/kernel`**. The `_com_raise_error` function you've made use of just happens to work, since there's no code path that goes through the final `throw` expression when a structured exception is raised. However, you can't count on C++-style stack unwinding to work when using SEH, except in very limited ways. And there's no guarantee that any of the other COM support functions will work at all. There's no intention of supporting the effect you've discovered under **`/kernel`**. It might quit working in the future. And, it won't be documented, because it's an unsupported use case.

Reading the other thread you referenced, I don't want you to get the idea that compiling with **`/kernel`** is some magic "get out of exceptions free" flag. SEH has always been supported both under **`/kernel`** and in user mode. If you can live with its restrictions on stack unwinding, you can use it in C++ code whether compiled under **`/kernel`** or not. Be aware, [these unwinding restrictions are substantial](https://docs.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=vs-2019#default-exception-handling-behavior), and not as robust as standard C++ unwinding semantics. SEH has the same kinds of costs as C++ exception handling, just smaller in magnitude without the type matching machinery and extra unwinding mechanics involved. [MSVC actually implements C++ exception handling on top of SEH internally](https://devblogs.microsoft.com/oldnewthing/20200117-00/?p=103338). Under **`/EHs`** it also does the extra stuff to guarantee standard C++ unwinding behavior. 

I hope this helps with understanding what's going on.

**Comment by DBJDBJ** [comment]
@corob-msft  Many thanks on time spent and information provided. I will spend some time investigating, quick points if I may:

- Terminology. When you say:  "get out of exceptions free". Do you mean C++ exceptions or SEH?
- Using MS STL under /kernel works. Am I right in saying this is not by accident? 
- `<comdef.h>` : I see no issue in making that "accidental" SEH modus operandi official? Or to put it differenlty: there is no point of tying `<comdef.h>` to C++ exceptions when MS STL is not?
- What is now the role of the `_CPPUNWIND` macro? It is not mentioned in [that good page](https://docs.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=vs-2019#default-exception-handling-behavior).
- All we/I do care is not to have WIN executables [bigger and slower due to usage of C++ exceptions](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1640r1.html). Not more and not less. 
- Humble I also use C99/C11/Win32 probably as a primary language. I am fully aware on what is SEH since approx 1994. (gasp!)

Again, many thanks for your time ... DBJ

**Comment by corob-msft** [comment]
Going point by point, I mean you don't need to use **`/kernel`** to create C++ code that has no standard C++ exception machinery. Compiling with **`/kernel`** is meant for kernel-mode code, which is used for drivers and OS components. Object files compiled for kernel mode may not play nicely when linked with user-mode code. If you're not building code to run in kernel mode, then I don't recommend you use that compiler option.

The parts of the STL that don't rely on exceptions do generally work in kernel mode. More of the STL was made kernel-mode compatible in the last major release. However, large sections of the STL rely on exception support, and those parts won't work.

There's no chance of <comdef.h> being sanctioned for kernel-mode code. COM utilities and the kernel are not designed or tested to work together. Using it may lead to blue screens, or worse. Consider it officially not recommended.

[`_CPPUNWIND`](https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=vs-2019) is a predefined preprocessor macro.. It's defined with a value of 1 if either **`/EHs`** or **`/EHa`** is specified, whether explicitly or implicitly, and undefined otherwise.

To avoid the overhead of C++ exceptions, don't define **`/EHs`** or **`/EHa`**. That's all you need to (not) do.

SEH dates back at least to the first Win32 APIs, so early 90's sounds about right. Sadly, I got rid of my reference books and software from that era twenty years ago, when they seemed hopelessly obsolete.

**Comment by DBJDBJ** [comment]
@corob-msft  ah ok thanks for warning about when and why not to use /kernel. '<comdef.h>' and friends I used last time commerically (I think) 1999 (gasp, again!), although it still works rather well. 

Honestly in all the projects spanning the decades, I never got that interested into anything elase but `/EHsc` . But now there is this ["clash of titans" ](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1947r0.pdf)in the C++ world arround exceptions. I am in the "no exceptions" corner (laugh), And I am OK with SEH. And I am not keen on any std lib in general.

MS STL seems to work for me with /kernel ?  Fully and completely including destructors calling.  Please look into [my toy project here](https://github.com/DBJDBJ/kernel_ms_stl).

There is only eight exceptions thrown out of MS STL (it seems). All [implemented here](https://github.com/microsoft/STL/blob/master/stl/src/xthrow.cpp).

In _NO_EXCEPTIONS builds, that  _THROW  macro is defined as
```cpp
// <yvals> #482
_THROW(x) x._Raise()
```
And in `<yvals.h>` there is this `_RAISE` macro which also seems to be used
```cpp
#ifdef _DEBUG
#define _RAISE(x) _invoke_watson(_CRT_WIDE(#x), __FUNCTIONW__, __FILEW__, __LINE__, 0)
#else // _DEBUG
#define _RAISE(x) _invoke_watson(nullptr, nullptr, nullptr, 0, 0)
#endif // _DEBUG
```
To cut to long story short, there is no official MS STL /kernel test being done. Ben Craig is [doing one "as we speak"](https://github.com/microsoft/STL/issues/1289#issuecomment-702276106). But I see no MS STL team involvement.

So far I need no test (lagh) [it all seem just to work for me](https://github.com/DBJDBJ/kernel_ms_stl) ...
